### PR TITLE
Queue preemptible work item

### PR DESCRIPTION
**Pull request by shpalani**
## Description

_Describe the purpose of and changes within this Pull Request._

## Testing

_Do any existing tests cover this change? Are new tests needed?_

## Documentation

_Is there any documentation impact for this change?_

### CODE REVIEW

File: libs/platform/ebpf_epoch.c [hunk 1]
```diff
@@ -495,6 +502,7 @@ ebpf_epoch_free_work_item(_Frees_ptr_opt_ ebpf_epoch_work_item_t* work_item)
     lock_state = ebpf_lock_lock(&_ebpf_epoch_cpu_table[current_cpu].lock);
     ebpf_list_remove_entry(&work_item->header.list_entry);
     ebpf_lock_unlock(&_ebpf_epoch_cpu_table[current_cpu].lock, lock_state);
+    ebpf_free_preemptible_work_item(work_item->preemptible_work_item);
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:505]: Actually, its not needed to call `ebpf_free_preemptible_work_item` here. The actual work item routine can free it.  Check this [MSDN article](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-iofreeworkitem).

File: libs/platform/user/ebpf_platform_user.cpp [hunk 2]
```diff
@@ -901,8 +901,6 @@ _ebpf_preemptible_routine(_Inout_ PTP_CALLBACK_INSTANCE instance, _In_opt_ void*
 
     ebpf_preemptible_work_item_t* work_item = (ebpf_preemptible_work_item_t*)parameter;
     work_item->work_item_routine(work_item->work_item_context);
-
-    ebpf_free_preemptible_work_item(work_item);
```
Comment (shankarseal) [inline libs/platform/user/ebpf_platform_user.cpp:905]: please revert.

Comment (shpalani) [reply libs/platform/user/ebpf_platform_user.cpp]: Sure, still needs investigation as it crashes ("Reason: Value unavailable error for BadAddress")here when added

`7e9ff5f0 00007ff6`a555e180     unit_tests!ebpf_free+0x48 [C:\ebpf\ebpf-for-windows\libs\platform\user\ebpf_platform_user.cpp @ 449] 
09 000000f9`7e9ff630 00007ff6`a555bcae     unit_tests!ebpf_free_preemptible_work_item+0x30 [C:\ebpf\ebpf-for-windows\libs\platform\user\ebpf_platform_user.cpp @ 917] 
0a 000000f9`7e9ff660 00007ff9`d6d5236a     unit_tests!_ebpf_preemptible_routine+0x5e [C:\ebpf\ebpf-for-windows\libs\platform\user\ebpf_platform_user.cpp @ 906]

File: libs/platform/ebpf_epoch.c (hunk 1 above)
Comment (shpalani) [reply libs/platform/ebpf_epoch.c]: Removed.

File: libs/platform/ebpf_epoch.c [hunk 3]
```diff
@@ -452,6 +453,12 @@ ebpf_epoch_allocate_work_item(_In_ void* callback_context, _In_ const void (*cal
...
     work_item->header.entry_type = EBPF_EPOCH_ALLOCATION_WORK_ITEM;
 
+    if (ebpf_allocate_preemptible_work_item(
+            &work_item->preemptible_work_item, work_item->callback, work_item->callback_context) != EBPF_SUCCESS) {
+        ebpf_free(work_item);
+        return NULL;
+    }
+
     return work_item;
 }
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:464]: @Alan-Jowett - is it the responsibility of the caller that allocated the `ebpf_epoch_work_item` to call `ebpf_epoch_free_work_item` from inside the `callback` ?

File: libs/platform/ebpf_epoch.c [hunk 4]
```diff
@@ -548,7 +555,7 @@ _ebpf_epoch_release_free_list(_Inout_ ebpf_epoch_cpu_entry_t* cpu_entry, int64_t
             break;
         case EBPF_EPOCH_ALLOCATION_WORK_ITEM: {
             ebpf_epoch_work_item_t* work_item = CONTAINING_RECORD(header, ebpf_epoch_work_item_t, header);
-            work_item->callback(work_item->callback_context);
+            ebpf_queue_preemptible_work_item(work_item->preemptible_work_item);
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:558]: Currently `_ebpf_preemptible_routine` calls `ebpf_free_preemptible_work_item` which frees `work_item_context`. Should it be the caller of `ebpf_allocate_preemptible_work_item` that should free that instead?

Comment (shpalani) [reply libs/platform/ebpf_epoch.c:558]: In this case, 
In ebpf_program_execute_test_run(), it has
   ebpf_allocate_preemptible_work_item(...) +  ebpf_queue_preemptible_work_item(work_item)
So, here caller cannot free the context, as ebpf_queue_preemptible_work_item API needs to complete to free.

**Comment by shpalani** [comment]
Abandoning this PR...

**Pull request closed without merging**